### PR TITLE
Rethinking En Passant Evasion Capture

### PR DESCRIPTION
**Pull request by BM123499**
STC https://tests.stockfishchess.org/tests/view/5ff890946019e097de3ef0a5
LLR: 2.93 (-2.94,2.94) {-1.25,0.25}
Total: 47912 W: 4327 L: 4285 D: 39300
Ptnml(0-2): 144, 3312, 17012, 3334, 154
bench: 4109336

Changing the way to check if En Passant Evasion Capture is pseudo-legal. No change on complexity, just checking if it is a discovery check instead of the checking piece is the double-pushed pawn.
The idea is to support a few more illegal fens without affecting legal fens. Solving issue #3270.

### CODE REVIEW

**Comment by BM123499** [comment]
@snicolet
a) it supports a few illegal fens (such as 8/8/8/1k6/3Pp3/8/8/4KQ2 b - d3 0 1) as shown in #3270.
b) there is no changing in `perft` numbers since it all moves listed in legal fens are the same as before. I tested `bench`, `bench 16 1 26` and some [predetermined positions](https://www.chessprogramming.org/Perft_Results), there was no changing in nodes searched.

There was a disclaimed that this changed slowed down the search by 1%, but there is no losing in elo at [Fishtest](https://tests.stockfishchess.org/tests/view/5ff890946019e097de3ef0a5).

In resume, there is no changing if you are looking at legal fens.

**Comment by ddobbelaere** [comment]
Although it "fixes" #3270 (essentially because more pseudo-legal moves are generated in the background, as a matter of fact), I find this PR highly confusing and have some doubts about it.

Let me try to rephrase what the current situation is w.r.t. master by considering a pawn push e2-e4 by white and considering black's legal moves (e3 is the e.p. square).

**master**: "An e.p. capture can only be a pseudo-legal move of type evasion if the pawn on e4 is a checking piece. Don't bother generating pseudo-legal moves if this is not the case."
**This PR**: "An e.p. capture can only be a pseudo-legal move of type evasion if e2 is not on a line of attack (checking line from slider) OR contains a checking piece (weird situation where there is a checking piece on e2)."

My question is: why is this e2 square so special? You might as well consider all e.p. captures as pseudo-legal evasions then...
If you want to filter out pseudo-legal moves beforehand, more logical would be: "An e.p. capture can only be a pseudo-legal move of type evasion if e4 is a checking piece **OR e3 is on a line of attack (or contains a checking piece, which is also pretty weird, but anyway, we are looking at "garden of Edens" here...)**. Don't bother generating pseudo-legal moves otherwise.", although naturally this increases code complexity...


TLDR: More pseudo-legal moves are being generated, which is why #3270 is "fixed", but the current code of this PR seems doubtful IMHO (I think it is equivalent in practice to just dropping the line `if (Type == EVASIONS && (target & (pos.ep_square() + Up)))` (and the line after it) because the position will almost never be true (also not in the example in #3270)).

**Comment by BM123499** [comment]
@ddobbelaere Changing the way it checks maybe seems confusing and generates more moves, but, as it was written:
> An en passant capture can be an evasion only if the checking piece
> is the double pushed pawn and so is in the target. Otherwise this
> is a discovery check and we are forced to do otherwise.
On legal positions, there are just 2 possibility for a En Passant capture be an evasion move, as shown, if it is a discovery check, you can't capture. If it is a discovery check, the pawn was covering the checking piece, in your example e2 (behind En Passant Square), so this is a important square to check.
It's more natural to add `or e3 in on line of sight`, but changing the perspective is the way that both support and don't compromise legal position.
It indeed generates more pseudo-legal moves, but just for illegal fens. It doesn't create more moves on legal position. Legally, there are just two possibility that the rules of the games limits it, other positions are considered `unreachable`, and so illegal. My change add En Passant moves to those illegal fen. So there are more pseudo-legal moves generated, but just for illegal fens.

I'm not fluent in English, so if I didn't answer your question properly or made myself unclear (even on the comment on the top of the code), you can re-ask me or suggest some changes.

**Comment by BM123499** [comment]
@ddobbelaere I just Tested how many times it triggers and how many it doesn't on a `bench` and on a `bench 16 1 26`, there is no difference.

**Comment by ddobbelaere** [comment]
@BM123499 Exactly! Hence why my suggestion of just dropping the check is superior w.r.t. this PR, as:

- Both are logically equivalent (same bench, same perft) for legal reachable positions.
- Mine is less instructions for legal reachable positions (condition is useless and will never fire) :)
- Mine allows more "gardens of Eden"...

**Comment by BM123499** [comment]
@ddobbelaere I tested this
`if (Type == EVASIONS && !(target & (pos.ep_square() - Up))){ ++EnPassantTest1; return moveList; } else if(Type == EVASIONS) ++EnPassantTest2;` and it showed the same results for master and for my PR, but you can test yours as well. if it's faster it will be even better than mine... :)

**Comment by ddobbelaere** [comment]
@ddobbelaere I feel stupid right now, but I now get the point why you consider the square e2 (for the reason that you state, namely that the double pawn push can reveal a discovered check) :)

I agree that the proposition in this PR is a possible solution for #3270, although my suggestion is also possible (simply dropping the check all together).

**Comment by BM123499** [comment]
@ddobbelaere no need to feel stupid, I knew this would confuse some people, since my PR checks a different condition instead of checking the opposite of master. What makes both to be the same logically is the rules of chess and not a generic logic check.

**Comment by ddobbelaere** [comment]
@BM123499 Minor nits: change "discovery check" in comment to "discovered check" (this is the correct term), change "doubled-pushed pawn" to "double pushed pawn" (in line with master and https://www.chessprogramming.org/En_passant)

File: src/movegen.cpp [hunk 1]
```diff
@@ -156,10 +156,10 @@ namespace {
         {
             assert(rank_of(pos.ep_square()) == relative_rank(Us, RANK_6));
 
-            // An en passant capture can be an evasion only if the checking piece
-            // is the double pushed pawn and so is in the target. Otherwise this
-            // is a discovery check and we are forced to do otherwise.
-            if (Type == EVASIONS && !(target & (pos.ep_square() - Up)))
+            // An en passant capture cannot be an evasion if it is a discovery check
+            // and so the square the doubled-pushed pawn came from is in the target.
+            // Otherwise, the checking piece is the double-pushed pawn and it is pseudo-legal.
```
Comment (ddobbelaere) [inline src/movegen.cpp:161]: Suggestion:

An en passant capture cannot be an evasion if there is a discovered check,
i.e. if the square the double pushed pawn came from is on an attacking line.
Otherwise, the checking piece is the double pushed pawn itself and the capture is pseudo-legal.

Comment (BM123499) [reply src/movegen.cpp:161]: Do I need to put _that is revealed by the double pawn push_? Because it already checked if there was a double pawn push and, for legal position, it must be a discovered check. Otherwise, black was already in check and it is a illegal position.

Comment (ddobbelaere) [reply src/movegen.cpp]: Yeah it's redundant for legal reachable position. Oth, it hints at why the start square of double pushed pawn is checked.

Either way is fine for me though...
